packages/eslint-plugin/src/rules/prefer-reduce-type-parameter.ts

```diff
@@ -0,0 +1,116 @@
+  }
+
+  if (
+    member.property.type === AST_NODE_TYPES.Literal &&
+    typeof member.property.value === 'string'
+  ) {
+    return member.property.value;
+  }
+
+  return null;
+};
+
+export default util.createRule({
+  name: 'prefer-reduce-type-parameter',
+  meta: {
+    type: 'problem',
+    docs: {
+      category: 'Best Practices',
+      recommended: 'error',
+      description:
+        'Prefer using type parameter when calling `Array#reduce` instead of casting',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      preferTypeParameter:
+        'Unnecessary cast: Array#reduce accepts a type parameter for the default value.',
+    },
+    fixable: 'code',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const service = util.getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    const reportPossibleArrayReduceCasting = (
+      node: TSESTree.CallExpression | TSESTree.OptionalCallExpression,
+    ): void => {
+      {
+        if (
+          !util.isMemberOrOptionalMemberExpression(node.callee) ||
+          getMemberExpressionName(node.callee) !== 'reduce'
+        ) {
+          return;
+        }
+
+        const [, secondArg] = node.arguments;
+
+        if (
+          node.arguments.length < 2 ||
+          (secondArg.type !== AST_NODE_TYPES.TSAsExpression &&
+            secondArg.type !== AST_NODE_TYPES.TSTypeAssertion)
+        ) {
+          return;
+        }
+
+        // Get the symbol of the `reduce` method.
+        const tsNode = service.esTreeNodeToTSNodeMap.get(node.callee);
+        const reduceSymbol = checker.getSymbolAtLocation(tsNode);
+        if (reduceSymbol == null) {
+          return;
+        }
+
+        // Check the owner type of the `reduce` method.
+        for (const methodDecl of reduceSymbol.declarations) {
+          const typeDecl = methodDecl.parent;
+          if (
+            ts.isInterfaceDeclaration(typeDecl) &&
+            ts.isSourceFile(typeDecl.parent) &&
+            typeDecl.name.escapedText === 'Array'
+          ) {
+            context.report({
+              messageId: 'preferTypeParameter',
+              node: secondArg,
+              fix: fixer => [
+                fixer.removeRange([
+                  secondArg.range[0],
+                  secondArg.expression.range[0],
+                ]),
+                fixer.removeRange([
+                  secondArg.expression.range[1],
+                  secondArg.range[1],
+                ]),
+                fixer.insertTextAfter(
+                  node.callee,
+                  `<${context
+                    .getSourceCode()
+                    .getText(secondArg.typeAnnotation)}>`,
+                ),
+              ],
+            });
+
+            return;
+          }
+        }
+      }
+    };
+
+    return {
+      OptionalCallExpression: reportPossibleArrayReduceCasting,
+      CallExpression: reportPossibleArrayReduceCasting,
```

Comment:
You could wrap up a bunch of your logic as an [eslint selector](https://eslint.org/docs/developer-guide/selectors)

```
':matches(CallExpression, OptionalCallExpression) > :matches(MemberExpression, OptionalMemberExpression).callee'
```

Reply:
Personally I prefer not using selectors for this kind of matching unless they save having to do expensive or complex checking, since TypeScript can't infer the results accurately.

However, I'm happy to follow your lead if you've got a preference you'd like followed :)

Reply:
The great thing about using selectors is that it causes your code to be called less often.
The tradeoff is that it is impossible to represent this in types without some form of codegen.

I think that it depends on how much "back and forth" you need to do with nodes.
If your rule takes a high level node, then digs deep into the AST to find a specific node, then that's probably better suited to a selector.

If your rule requires it to track multiple nodes at various positions in the AST, then it's probably okay to not use a selector.